### PR TITLE
[DIAG-944] Batch Mode

### PR DESCRIPTION
**Pull request by gdj0nes**
Batch mode allows you to specify specific data rows rather than entire datasets for labeling

https://docs.labelbox.com/docs/batch-queue

* First version of the batch mode in the SDK

### CODE REVIEW

File: labelbox/schema/project.py [hunk 1]
```diff
@@ -424,6 +430,106 @@ def setup(self, labeling_frontend, labeling_frontend_options):
...
+        """
+        if mode == "BATCH":
+            self._update_queue_mode("ENABLED")
+
+        elif mode == "DATASET":
+            self._update_queue_mode("DISABLED")
+        else:
+            raise ValueError(
+                "Must provide either `BATCH` or `DATASET` as a mode")
+
```
Comment (gdj0nes) [inline labelbox/schema/project.py:499]: @msokoloff1 how hard would it be to get this into the update call

Comment (gdj0nes) [reply labelbox/schema/project.py]: i added it but it might be a bit of a hack

File: labelbox/data/annotation_types/data/raster.py [hunk 2]
```diff
@@ -51,6 +51,16 @@ def from_2D_arr(cls, arr: Union[TypedArray[Literal['uint8']],
...
 
+    @classmethod
+    def from_2D_array(cls, arr: TypedArray[Literal['uint8']], **kwargs):
+
+        if len(arr.shape) != 2:
+            raise ValueError(
+                f"Found array with shape {arr.shape}. Expected two dimensions ([W,H])"
+            )
+        arr = np.stack((arr,) * 3, axis=-1)
+        return cls(arr=arr, **kwargs)
```
Comment (msokoloff1) [inline labelbox/data/annotation_types/data/raster.py:62]: I think this is something old you might have had locally.

File: labelbox/schema/project.py [hunk 3]
```diff
@@ -424,6 +443,99 @@ def setup(self, labeling_frontend, labeling_frontend_options):
...
+        if self._is_batch_mode():
+            return QueueMode.Batch
+        else:
+            return QueueMode.Dataset
+
+    def _is_batch_mode(self):
+        if self.tag_set_status == "ENABLED":
+            return True
+        else:
+            return False
```
Comment (msokoloff1) [inline labelbox/schema/project.py:537]: return self.tag_set_status == "ENABLED"

File: labelbox/schema/project.py [hunk 4]
```diff
@@ -424,6 +443,99 @@ def setup(self, labeling_frontend, labeling_frontend_options):
...
+                __typename
+            }
+        }       
+        """ % "setTagSetStatusPyApi"
+
+        self.client.execute(query_str, {
+            'projectId': self.uid,
+            'status': status
+        })
+        self.tag_set_status = status
```
Comment (msokoloff1) [inline labelbox/schema/project.py:524]: tag_set_status is not going to be defined until this point. On Project init we should fetch the existing status.

File: labelbox/schema/project.py [hunk 5]
```diff
@@ -424,6 +443,99 @@ def setup(self, labeling_frontend, labeling_frontend_options):
...
+        if len(data_rows) > 0:
+            warnings.warn("Some Data Rows were not submitted successfully")
+
+        return res
+
+    def _update_queue_mode(self, mode: QueueMode):
+
+        if mode == QueueMode.Batch:
+            status = "ENABLED"
+        elif mode == QueueMode:
```
Comment (msokoloff1) [inline labelbox/schema/project.py:501]: QueueMode .Dataset

File: labelbox/schema/project.py [hunk 6]
```diff
@@ -424,6 +443,99 @@ def setup(self, labeling_frontend, labeling_frontend_options):
...
+                }
+                __typename
+            }
+        }       
+        """ % "setTagSetStatusPyApi"
+
+        self.client.execute(query_str, {
+            'projectId': self.uid,
+            'status': status
+        })
```
Comment (msokoloff1) [inline labelbox/schema/project.py:523]: Maybe not run this if status == self.tag_set_status

File: labelbox/schema/project.py [hunk 7]
```diff
@@ -424,6 +443,99 @@ def setup(self, labeling_frontend, labeling_frontend_options):
...
+
+    def queue_mode(self):
+
+        if self._is_batch_mode():
+            return QueueMode.Batch
+        else:
+            return QueueMode.Dataset
+
+    def _is_batch_mode(self):
+        if self.tag_set_status == "ENABLED":
```
Comment (msokoloff1) [inline labelbox/schema/project.py:534]: We also should just remove `tag_set_status` from this completely. We can track this via a queue_mode class variable that can be set on init. This way queue mode will be consistent with how other fields are updated in the SDK.

File: labelbox/schema/project.py [hunk 8]
```diff
@@ -424,6 +443,99 @@ def setup(self, labeling_frontend, labeling_frontend_options):
...
+        method = "removeBatchOfDataRows"
+        return self._post_batch(method, data_rows)
+
+    def _post_batch(self, method, data_rows):
+        """Create """
+
+        ids = [dr.uid for dr in data_rows]
+
+        if len(ids) > MAX_BATCH_SIZE:
+            raise ValueError("1000 Max DataRows at a time")
```
Comment (msokoloff1) [inline labelbox/schema/project.py:469]: Use MAX_BATCH_SIZE in the error

File: labelbox/schema/project.py [hunk 9]
```diff
@@ -79,6 +90,14 @@ class Project(DbObject, Updateable, Deletable):
     benchmarks = Relationship.ToMany("Benchmark", False)
     ontology = Relationship.ToOne("Ontology", True)
 
+    def update(self, **kwargs):
```
Comment (msokoloff1) [inline labelbox/schema/project.py:93]: Is there any danger to toggling the mode? What happens to existing annotations or data in the queue?

Comment (gdj0nes) [reply labelbox/schema/project.py:99]: will be handled by product in the future

File: labelbox/schema/project.py (hunk 8 above)
Comment (gdj0nes) [reply labelbox/schema/project.py:473]: done

**Comment by GrantEating** [comment]
> * First version of the batch mode in the SDK

Can you provide a more lengthy description? I dont know what batch mode is. I would love to understand the ideas behind the PR before reviewing

File: labelbox/schema/project.py [hunk 10]
```diff
@@ -424,6 +449,108 @@ def setup(self, labeling_frontend, labeling_frontend_options):
...
+            raise ValueError("Project must be in batch mode")
+
+        method = "removeBatchOfDataRows"
+        return self._post_batch(method, data_row_ids)
+
+    def _post_batch(self, method, data_row_ids: List[str]):
+        """Create """
+
+        if len(data_row_ids) > MAX_BATCH_SIZE:
+            raise ValueError(f"Exceed max batch size of {MAX_BATCH_SIZE}")
```
Comment (msokoloff1) [inline labelbox/schema/project.py:473]: Exceeded

File: tests/integration/test_project.py [hunk 11]
```diff
@@ -107,3 +106,10 @@ def test_attach_instructions(client, project):
 def test_queued_data_row_export(configured_project):
     result = configured_project.export_queued_data_rows()
     assert len(result) == 1
+
+
+def test_queue_mode(configured_project: Project):
+    assert configured_project.queue_mode() == QueueMode.Dataset
+    configured_project.update(queue_mode=QueueMode.Batch)
+    # TODO: understand why this fails
```
Comment (msokoloff1) [inline tests/integration/test_project.py:114]: Remove

File: tests/data/annotation_types/geometry/test_rectangle.py [hunk 12]
```diff
@@ -16,5 +15,9 @@ def test_rectangle():
     expected['coordinates'] = tuple([tuple([tuple(x) for x in points[0]])])
     assert rectangle.shapely.__geo_interface__ == expected
 
-    raster = rectangle.draw(height=32, width=32)
-    assert (cv2.imread("tests/data/assets/rectangle.png") == raster).all()
+    # raster = rectangle.draw(height=32, width=32)
+    # assert (cv2.imread("tests/data/assets/rectangle.png") == raster).all()
```
Comment (msokoloff1) [inline tests/data/annotation_types/geometry/test_rectangle.py:19]: What happened here?

File: labelbox/schema/project.py [hunk 13]
```diff
@@ -424,6 +449,108 @@ def setup(self, labeling_frontend, labeling_frontend_options):
...
+
+        return mode
+
+    def queue_mode(self):
+
+        query_str = """query %s($projectId: ID!) {
+              project(where: {id: $projectId}) {
+                 id
+                 tagSetStatus
+                __typename
```
Comment (msokoloff1) [inline labelbox/schema/project.py:537]: you don't need __typename or id

**Review by msokoloff1 — CHANGES_REQUESTED** [review]
just a couple of small things

File: labelbox/schema/project.py [hunk 14]
```diff
@@ -424,6 +449,106 @@ def setup(self, labeling_frontend, labeling_frontend_options):
...
+                "Must provide either `BATCH` or `DATASET` as a mode")
+
+        query_str = """mutation %s($projectId: ID!, $status: TagSetStatusInput!) {
+              project(where: {id: $projectId}) {
+                 setTagSetStatus(input: {tagSetStatus: $status}) {
+                    id
+                    tagSetStatus
+                    __typename
+                }
+                __typename
```
Comment (msokoloff1) [inline labelbox/schema/project.py:519]: you don't need the return types

File: tests/data/annotation_types/geometry/test_rectangle.py (hunk 12 above)
Comment (gdj0nes) [reply tests/data/annotation_types/geometry/test_rectangle.py:19]: unsure but I'll uncommit to see if it passes

**Review by msokoloff1 — APPROVED** [review]

**Pull request merged**